packages/twoside/init.lua

```diff
@@ -65,28 +66,74 @@ function package:_init (options)
   self.class:registerPostinit(function (class)
     class:mirrorMaster(options.oddPageMaster, options.evenPageMaster)
   end)
+  self.class:registerHook("newpage", spreadHook)
   self.class:registerHook("newpage", switchPage)
 end
 
 function package:registerCommands ()
 
   self:registerCommand("open-double-page", function()
-    SILE.typesetter:leaveHmode()
+    SILE.call("open-spread", { double = false, odd = true, blank = false })
+  end)
+
+  self:registerCommand("open-spread-eject", function()
     SILE.call("supereject")
-    if self.class:oddPage() then
-      SILE.typesetter:typeset("")
-      SILE.typesetter:leaveHmode()
-      SILE.call("supereject")
+  end)
+
+  -- This is upstreamed from CaSILE. Similar to the original open-double-page,
+  -- but can disable headers and folios on blank pages and allows opening the
+  -- even side (with or without a leading blank).
+  self:registerCommand("open-spread", function (options)
+    local odd = SU.boolean(options.odd, true)
+    local double = SU.boolean(options.double, true)
+    local blank = SU.boolean(options.blank, true)
+    local optionsMet = function ()
+      return (not double or spread_counter > 1) and
+             (odd and self.class:oddPage()) or
+             (not odd and not self.class:oddPage())
     end
+    spread_counter = 0
     SILE.typesetter:leaveHmode()
+    -- Output a box, then remove it otherwise we can't prove what page new content will land on
+    SILE.call("hbox"); SILE.typesetter:leaveHmode(); table.remove(SILE.typesetter.state.nodes)
+    if spread_counter == 1 and optionsMet() then
+      SILE.typesetter.state.outputQueue = {}
```

Comment:
Why do we need this? Can we investigate what's left in the outputQueue (vglues, perhaps) and document the rationale?

Reply:
I've expanded the comment to explain. This is for an edge case where the page is busting at the seams already but `:leaveHmode()` won't trigger a page break because of widow penalties. Adding the hbox here *after* having left hmode forces a line to be added and hence the next tme we leave hmode *will* break the page. For this edge case the *only* thing in the output queue will be the extra line we just added.